File: billops-backend/app/schemas/billing_rule.py
from datetime import datetime
from uuid import UUID
from decimal import Decimal
from pydantic import BaseModel, Field


class BillingRuleCreate(BaseModel):
    """Schema for creating a new billing rule."""
    project_id: UUID
    rule_type: str = Field(..., min_length=1)  # hourly | fixed | retainer
    rate_cents: int = Field(..., ge=0)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    rounding_increment_minutes: int | None = None
    overtime_multiplier: Decimal | None = None
    cap_hours: Decimal | None = None
    retainer_hours: Decimal | None = None
    effective_from: datetime | None = None
    effective_to: datetime | None = None
    meta: dict | None = None


class BillingRuleUpdate(BaseModel):
    """Schema for updating a billing rule."""
    rule_type: str | None = None
    rate_cents: int | None = None
    currency: str | None = None
    rounding_increment_minutes: int | None = None
    overtime_multiplier: Decimal | None = None
    cap_hours: Decimal | None = None
    retainer_hours: Decimal | None = None
    effective_from: datetime | None = None
    effective_to: datetime | None = None
    meta: dict | None = None


class BillingRuleResponse(BaseModel):
    """Schema for billing rule response."""
    id: UUID
    project_id: UUID
    rule_type: str
    rate_cents: int
    currency: str
    rounding_increment_minutes: int | None
    overtime_multiplier: Decimal | None
    cap_hours: Decimal | None
    retainer_hours: Decimal | None
    effective_from: datetime
    effective_to: datetime | None
    meta: dict | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: billops-frontend/src/schemas/billingRule.ts
import { z } from 'zod';
import type { RuleType } from '@/config/billingRuleConfig';

export const billingRuleFormSchema = z.object({
  project_id: z.string().uuid('Project is required'),
  rule_type: z.string().min(1, 'Rule type is required') as z.ZodType<RuleType>,
  rate_cents: z.coerce.number().min(0, 'Rate must be 0 or greater'),
  currency: z.string().length(3, 'Currency must be 3 letters').default('USD'),
  
  // Hourly rule fields
  rounding_increment_minutes: z.coerce.number().optional(),
  overtime_multiplier: z.coerce.number().optional(),
  cap_hours: z.coerce.number().optional(),
  
  // Retainer rule fields
  retainer_hours: z.coerce.number().optional(),
  
  // Session-based rule fields
  session_max_duration: z.coerce.number().optional(),
  
  // Per-item rule fields
  item_description: z.string().optional(),
  
  // Tiered rule fields
  tier1_threshold: z.coerce.number().optional(),
  tier1_rate_cents: z.coerce.number().optional(),
  tier2_threshold: z.coerce.number().optional(),
  tier2_rate_cents: z.coerce.number().optional(),
  tier3_rate_cents: z.coerce.number().optional(),
  
  // Event-based rule fields
  event_types: z.array(z.string()).optional(),
  rate_per_event_cents: z.coerce.number().optional(),
  
  // Travel fee fields
  hourly_rate_cents: z.coerce.number().optional(),
  mileage_rate_cents: z.coerce.number().optional(),
  parking_charges: z.boolean().optional(),
  
  // Date fields
  effective_from: z.string().optional(),
  effective_to: z.string().optional(),
  
  // Metadata
  meta: z.record(z.any()).optional(),
}).refine(
  (data) => {
    if (!data.effective_from || !data.effective_to) return true;
    return new Date(data.effective_from) < new Date(data.effective_to);
  },
  {
    message: 'Effective end date must be after start date',
    path: ['effective_to'],
  }
).refine(
  (data) => {
    if (data.rule_type === 'retainer') {
      return data.retainer_hours && data.retainer_hours > 0;
    }
    return true;
  },
  {
    message: 'Retainer rules must specify included hours',
    path: ['retainer_hours'],
  }
).refine(
  (data) => {
    if (data.rule_type === 'tiered') {
      return (
        (data.tier1_rate_cents ?? 0) >= 0 &&
        (data.tier2_rate_cents ?? 0) >= 0
      );
    }
    return true;
  },
  {
    message: 'Tiered pricing must have tier rates defined',
    path: ['tier1_rate_cents'],
  }
);

export type BillingRuleFormData = z.infer<typeof billingRuleFormSchema>;
